SimPEG/utils/model_utils.py

```diff
@@ -258,17 +263,45 @@ def distance_weighting(
 
     reference_locs = np.asarray(reference_locs)
 
-    # Calculate distance from receiver locations
-    # reference_locs is a scalar
-    if reference_locs.ndim < 2:
-        distance = np.abs(mesh.cell_centers[:, -1] - reference_locs)
+    cell_centers = mesh.cell_centers[active_cells]
+    cell_volumes = mesh.cell_volumes[active_cells]
+
+    # address 1D case
+    if mesh.dim == 1:
+        cell_centers = cell_centers.reshape(-1, 1)
+        reference_locs = reference_locs.reshape(-1, 1)
+
+    if engine == "numba":
+
+        @njit(parallel=True)
+        def distance_weighting_numba(
+            cell_centers: np.ndarray,
+            cell_volumes: np.ndarray,
+            reference_locs: np.ndarray,
+            threshold: float,
+            exponent: float = 2.0,
+        ):
+            distance_weights = np.zeros(len(cell_centers))
+            for _, rl in enumerate(reference_locs):
+                dst_wgt = (
+                    np.sqrt(((cell_centers - rl) ** 2).sum(axis=1)) + threshold
+                ) ** exponent
+                dst_wgt = (cell_volumes / dst_wgt) ** 2
+                distance_weights += dst_wgt
+
+            return distance_weights
```

Comment:
I would move this function definition outside of the public function, for example move it to 
```python
@njit(parallel=True)
def _distance_weighting_numba(...):
    ...
```

By defining numba decorated functions within another function body, the interpreter will compile it when it needs to run it, but as soon as we leave the scope of `distance_weighting`, it will throw the compiled bits to the garbage collector. Therefore, every time a user runs this function, it will have to recompile the Numba function.

On the other hand, if it lives outside of it, it will compile it only the first time it needs to use it. Then, any further call to `distance_weighting` will reuse those compiled bits.